### ISSUE TITLE
make diskless instance migrateable

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#293]
Originally reported of Google Code with ID 237.

```
What software version are you running? Please provide the output of "gnt-
cluster --version" and "gnt-cluster version".

gnt-cluster (ganeti v2.5.0) 2.5.0

Software version: 2.5.0
Internode protocol: 2050000
Configuration format: 2050000
OS api version: 20
Export interface: 0

<b>What distribution are you using?</b>

debian 6.0.4

<b>What steps will reproduce the problem?</b>
$ gnt-instance add --no-name-check --no-ip-check --no-install -t diskless -n some-node -o debootstrap+default testvm
# gnt-instance migrate --force -n other-node testvm 

<b>What is the expected output? What do you see instead?</b>

diskless should behave like blockdev/sharedfile, but:

Failure: prerequisites not met for this operation:
("Instance's disk layout 'diskless' does not allow migrations", 'wrong_state')

TIA Sascha.
```
Originally added on 2012-05-11 12:48:20 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Makes sense, indeed, thanks for reporting.
```
Originally added on 2012-05-11 12:57:39 +0000 UTC.

Changed State: Accepted

**Comment by Ganeti-Issues-Migrator** [comment]
```
I ran into the same problem and "fixed" this by adding the disk template 'DT_DISKLESS' to the set of externally-mirrored disk templates 'DTS_EXT_MIRROR'.
This is btw in our case kind of true because we are not having diskless instances, but instances booting from iSCSI, initialized by DHCP.

--- a/lib/constants.py  2012-04-30 17:43:01.735381572 +0000
+++ b/lib/constants.py  2012-04-30 17:42:23.150009051 +0000
@@ -396,7 +396,7 @@
 DTS_INT_MIRROR = frozenset([DT_DRBD8])
 
 # the set of externally-mirrored disk templates (e.g. SAN, NAS)
-DTS_EXT_MIRROR = frozenset([DT_SHARED_FILE, DT_BLOCK])
+DTS_EXT_MIRROR = frozenset([DT_SHARED_FILE, DT_BLOCK, DT_DISKLESS])
 
 # the set of non-lvm-based disk templates
 DTS_NOT_LVM = frozenset([DT_DISKLESS, DT_FILE, DT_SHARED_FILE, DT_BLOCK])

On this 'diskless' instances you only have a primary node and no secondary. for migration you have to give 'gnt-instance migrate' the '--node' flag.

in the rapi client the node flag is missing. you can add it by applying the patch:

--- original/client.py  2012-05-14 10:13:42.000000000 +0200
+++ new/client.py       2012-05-14 11:02:29.000000000 +0200
@@ -1049,7 +1049,7 @@
                              ("/%s/instances/%s/export" %
                               (GANETI_RAPI_VERSION, instance)), None, body)
 
-  def MigrateInstance(self, instance, mode=None, cleanup=None):
+  def MigrateInstance(self, instance, mode=None, cleanup=None, target_node=None):
     """Migrates an instance.
 
     @type instance: string
@@ -1058,6 +1058,8 @@
     @param mode: Migration mode
     @type cleanup: bool
     @param cleanup: Whether to clean up a previously failed migration
+    @type target_node: string
+    @param target_node: Target Node for migration
     @rtype: string
     @return: job id
 
@@ -1070,6 +1072,9 @@
     if cleanup is not None:
       body["cleanup"] = cleanup
 
+    if target_node is not None:
+      body["target_node"] = target_node
+
     return self._SendRequest(HTTP_PUT,
                              ("/%s/instances/%s/migrate" %
                               (GANETI_RAPI_VERSION, instance)), None, body)

with that approach i have not discovered a way to use the htools iallocator 'hail'. hail is needing a disk on the nodes and on the instances as well:

root@node:~# hscan 
Name  Nodes  Inst BNode BInst  t_mem  f_mem t_disk f_disk      Score

Error for LOCAL: failed to load data. Details:
Node 'node', error while reading attribute 'dtotal': Cannot convert value 'JSNull', error: Unable to read Double

i hope this might be helpful.

danielk
```
Originally added on 2012-05-21 19:29:00 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2012-07-19 15:29:55 +0000 UTC.

Added Labels: Type-Enhancement Component-htools

**Comment by Ganeti-Issues-Migrator** [comment]
```
with these two additional patches against backend.py and cmdlib.py i am able to use 'hail' in a diskless setup:

diff -u /usr/share/pyshared/ganeti/backend.py.20120920 /usr/share/pyshared/ganeti/backend.py
--- /usr/share/pyshared/ganeti/backend.py.20120920      2012-09-20 14:02:48.618436770 +0200
+++ /usr/share/pyshared/ganeti/backend.py       2012-09-20 14:03:21.628336950 +0200
@@ -532,8 +532,8 @@
     vg_free = int(round(vginfo[0][0], 0))
     vg_size = int(round(vginfo[0][1], 0))
   else:
-    vg_free = None
-    vg_size = None
+    vg_free = 0
+    vg_size = 0
 
   return {
     "name": name,

diff -u /usr/share/pyshared/ganeti/cmdlib.py.20120920 /usr/share/pyshared/ganeti/cmdlib.py
--- /usr/share/pyshared/ganeti/cmdlib.py.20120920       2012-09-20 14:46:52.700850651 +0200
+++ /usr/share/pyshared/ganeti/cmdlib.py        2012-09-20 14:47:43.880717301 +0200
@@ -9082,7 +9082,7 @@
   """
   # Required free disk space as a function of disk and swap space
   req_size_dict = {
-    constants.DT_DISKLESS: None,
+    constants.DT_DISKLESS: 0,
     constants.DT_PLAIN: sum(d[constants.IDISK_SIZE] for d in disks),
     # 128 MB are added for drbd metadata for each disk
     constants.DT_DRBD8:

htools are capable of collecting node data now. the hbal algorithm on the other hand is not working properly. the placement restrictions using 'hail' are working but the balancing part is not finding any solutions. i think this is because in my diskless setup i am only having primary nodes but no secondary nodes.

any hints on tweaking the htools for balancing diskless instances ?

danielk
```
Originally added on 2012-09-21 07:56:52 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
This will be part of 2.7.0 beta2. Note that for now explicit failover has been tested, but the use of a iallocator has not.

Thanks,

Guido

commit d8f0b7c197cb17423a94cca4ab6385fea534f715
Author: Guido Trotter <ultrotter@google.com>
Date:   Tue Mar 12 16:07:04 2013 +0100

    Make diskless instances externally mirrored
    
    This addresses Issue 237.
    
    Mirroring no disk is a no-op. As such we can treat them like mirrored
    instances, since the data they need (none) will be present on all nodes.
    
    This is definitely enough to failover or migrate instances with a manual
    target node. Further work might be needed for support in hail.
    
    Signed-off-by: Guido Trotter <ultrotter@google.com>
    Reviewed-by: Iustin Pop <iustin@google.com>

```
Originally added on 2013-03-12 17:09:25 +0000 UTC.

Mentioned Issue with GoogleCode ID 237 has been migrated to GitHub with ID #293.
Changed State: Started

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2013-03-12 17:09:41 +0000 UTC.

Added to Milestone: Release2.7

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2013-04-04 09:19:10 +0000 UTC.

Changed State: Fixed

**Comment by Ganeti-Issues-Migrator** [comment]
```
Marking 2.7 bugs fixed, as 2.7.0~rc1 was released, and should be ready to be tested in production environments.
```
Originally added on 2013-05-09 05:42:48 +0000 UTC.

Changed State: Released
